### PR TITLE
xtensa-build-zephyr: add imx

### PR DESCRIPTION
**Pull request by marc-hb**
Building in CI is required to test changes like #4524 and others

### CODE REVIEW

**Comment by marc-hb** [comment]
IMX successfully built in https://github.com/thesofproject/sof/pull/4525/checks?check_run_id=3130170363

File: scripts/xtensa-build-zephyr.sh [hunk 1]
```diff
@@ -89,11 +99,16 @@ build()
...
 	mkdir -p ${STAGING}/sof/ # smex does not use 'install -D'
 
+        # Default RIMAGE_KEY defined at the top of this file
 	for platform in "${PLATFORMS[@]}"; do
 		case "$platform" in
 			apl)
 				PLAT_CONFIG='intel_adsp_cavs15'
 				;;
+			imx)
+				PLAT_CONFIG='intel_adsp_cavs15'
```
Comment (marc-hb) [inline scripts/xtensa-build-zephyr.sh:109]: Hasty copy/paste... @iuliana-prodan which `--board` parameter have you been using?

File: scripts/xtensa-build-zephyr.sh [hunk 2]
```diff
@@ -7,7 +7,7 @@ set -e
 
 SOF_TOP=$(cd "$(dirname "$0")" && cd .. && pwd)
 
-SUPPORTED_PLATFORMS=(apl cnl icl tgl-h)
+SUPPORTED_PLATFORMS=(apl cnl icl tgl-h imx)
 # Default value, can (and sometimes must) be overridden with -p
```
Comment (dbaluta) [inline scripts/xtensa-build-zephyr.sh:11]: Shouldn't we add imx8 instead of imx here?

File: scripts/xtensa-build-zephyr.sh [hunk 3]
```diff
@@ -140,10 +145,12 @@ build()
 			       -l "$STAGING"/sof/sof-"$platform".ldc \
 			       "$bdir"/zephyr/zephyr.elf
 
+                        test "$RIMAGE_KEY" = 'None' ||
 			west sign  --build-dir "$bdir" \
```
Comment (dbaluta) [inline scripts/xtensa-build-zephyr.sh:149]: Indentation looks odd here. Same with line below.

File: scripts/xtensa-build-zephyr.sh (hunk 1 above)
Comment (paulstelian97) [reply scripts/xtensa-build-zephyr.sh:109]: I believe it's nxp_adsp_imx8.

File: scripts/xtensa-build-zephyr.sh (hunk 3 above)
Comment (paulstelian97) [reply scripts/xtensa-build-zephyr.sh:149]: Tabs vs spaces, the new line seems to be indented with spaces.

**Comment by iuliana-prodan** [comment]
@marc-hb I already have a patch for this, but is only local...since Zephyr pull request is not yet merged - see https://github.com/zephyrproject-rtos/zephyr/pull/35467

Here's my patch:
```
 scripts/xtensa-build-zephyr.sh | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/scripts/xtensa-build-zephyr.sh b/scripts/xtensa-build-zephyr.sh
index ce27053d2..6016a49ad 100755
--- a/scripts/xtensa-build-zephyr.sh
+++ b/scripts/xtensa-build-zephyr.sh
@@ -7,7 +7,7 @@ set -e
 
 SOF_TOP=$(cd "$(dirname "$0")" && cd .. && pwd)
 
-SUPPORTED_PLATFORMS=(apl cnl icl tgl-h)
+SUPPORTED_PLATFORMS=(apl cnl icl tgl-h imx8)
 # Default value, can be overridden on the command line
 WEST_TOP="${SOF_TOP}"/zephyrproject
 BUILD_JOBS=$(nproc --all)
@@ -104,6 +104,10 @@ build()
 				PLAT_CONFIG='intel_adsp_cavs25'
 				RIMAGE_KEY=modules/audio/sof/keys/otc_private_key_3k.pem
 				;;
+                        imx8)
+                                PLAT_CONFIG='nxp_adsp_imx8'
+                                ;;
+
 			*)
 				echo "Unsupported platform: $platform"
 				exit 1
-- 
2.17.1
```
With this you compile imx8 with: 
`zephyrproject$ ./modules/audio/sof/scripts/xtensa-build-zephyr.sh -p . imx8`

The commands for west are:
```
zephyrproject$ west build -d imx -b nxp_adsp_imx8 zephyr/samples/subsys/audio/sof/
zephyrproject$ west sign -d imx -p rimage/build/rimage -t rimage -D rimage/config -- -k no
```

File: scripts/xtensa-build-zephyr.sh (hunk 1 above)
Comment (iuliana-prodan) [reply scripts/xtensa-build-zephyr.sh:109]: yes, is ` PLAT_CONFIG='nxp_adsp_imx8'`

File: scripts/xtensa-build-zephyr.sh (hunk 2 above)
Comment (iuliana-prodan) [reply scripts/xtensa-build-zephyr.sh:11]: yes, should be imx8:
```
-SUPPORTED_PLATFORMS=(apl cnl icl tgl-h)
+SUPPORTED_PLATFORMS=(apl cnl icl tgl-h imx8)
```

**Comment by marc-hb** [comment]
> @marc-hb I already have a patch for this, but is only local...since Zephyr pull request is not yet merged - see zephyrproject-rtos/zephyr#35467

Thanks this helps. You can already submit the `PLAT_CONFIG='nxp_adsp_imx8'` right now, that takes away the guesswork and makes your local diffs (and others' including mine!) smaller. As long as you don't change the `SUPPORTED_PLATFORMS` yet it won't get built by default. For the icing on the cake you could add a WIP link to zephyrproject-rtos/zephyr#35467 next to `PLAT_CONFIG='nxp_adsp_imx8'`.

BTW what about the RIMAGE_KEY? You don't care for now that this script signs sof-imx with the Intel key? Just curious.

So I cherry-picked zephyrproject-rtos/zephyr#35467 but now I'm gettting this with the Zephyr SDK 0.12.2, any clue?
```
zephyr/arch/xtensa/core/offsets/offsets.c
In file included from zephyr/include/arch/xtensa/arch.h:27,
                 from zephyr/include/arch/cpu.h:27,
                 from zephyr/include/kernel_includes.h:33,
                 from zephyr/include/kernel.h:17,
                 from zephyr/include/init.h:11,
                 from zephyr/include/device.h:29,
                 from zephyr/kernel/include/kernel_offsets.h:6,
                 from zephyr/arch/xtensa/core/offsets/offsets.c:7:
zephyr/include/arch/xtensa/irq.h:10:10: fatal error: xtensa/config/core-isa.h: No such file or directory
   10 | #include <xtensa/config/core-isa.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~
      
```

**Comment by iuliana-prodan** [comment]
> > @marc-hb I already have a patch for this, but is only local...since Zephyr pull request is not yet merged - see [zephyrproject-rtos/zephyr#35467](https://github.com/zephyrproject-rtos/zephyr/pull/35467)
> 
> Thanks this helps. You can already submit the `PLAT_CONFIG='nxp_adsp_imx8'` right now, that takes away the guesswork and makes your local diffs (and others' including mine!) smaller. As long as you don't change the `SUPPORTED_PLATFORMS` yet it won't get built by default. For the icing on the cake you could add a WIP link to [zephyrproject-rtos/zephyr#35467](https://github.com/zephyrproject-rtos/zephyr/pull/35467) next to `PLAT_CONFIG='nxp_adsp_imx8'`.
> 
Ok, I will make a draft PR for adding `PLAT_CONFIG='nxp_adsp_imx8'` in `xtensa_build_zephyr.sh`.

> BTW what about the RIMAGE_KEY? You don't care for now that this script signs sof-imx with the Intel key? Just curious.
> 
TBH I thought that if I don't specify a RIMAGE_KEY [here](https://github.com/thesofproject/sof/blob/main/scripts/xtensa-build-zephyr.sh#L106), the .ri image is not signed. I'll have to double check this.

> So I cherry-picked [zephyrproject-rtos/zephyr#35467](https://github.com/zephyrproject-rtos/zephyr/pull/35467) and #4524 but now I'm gettting this with the Zephyr SDK 0.12.2, any clue?
> 
> ```
> zephyr/arch/xtensa/core/offsets/offsets.c
> In file included from zephyr/include/arch/xtensa/arch.h:27,
>                  from zephyr/include/arch/cpu.h:27,
>                  from zephyr/include/kernel_includes.h:33,
>                  from zephyr/include/kernel.h:17,
>                  from zephyr/include/init.h:11,
>                  from zephyr/include/device.h:29,
>                  from zephyr/kernel/include/kernel_offsets.h:6,
>                  from zephyr/arch/xtensa/core/offsets/offsets.c:7:
> zephyr/include/arch/xtensa/irq.h:10:10: fatal error: xtensa/config/core-isa.h: No such file or directory
>    10 | #include <xtensa/config/core-isa.h>
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~
>       
> ```
For this you'll have to update the xtensa_hal, also. I had a PR on that repo, also. See [here](https://github.com/zephyrproject-rtos/hal_xtensa/commits/master).
I see that `xtensa_build_zephyr.sh` is doing `west update zephyr hal_xtensa`, but maybe is not taking the latest changes?
Check if you have the latest commit from hal_xtensa.

**Comment by marc-hb** [comment]
> TBH I thought that if I don't specify a RIMAGE_KEY here, the .ri image is not signed. I'll have to double check this.

No there is a default value at the top. I don't think the build can produce any unsigned file with `.ri` extension right now (which would be confusing).

I was confused by this default value too which is why I added a comment in this PR. Maybe the default value should simply be moved down next to the relevant code.

> For this you'll have to update the xtensa_hal, also. I had a PR on that repo, also. See here.
> I see that xtensa_build_zephyr.sh is doing west update zephyr hal_xtensa, but maybe is not taking the latest changes?

It won't because you didn't update the west manifest in the Zephyr PR. If you want your Zephyr PR to update hal_extensa there must be a west manifest update in your Zephyr PR pointing at the new hal extensa version.

https://docs.zephyrproject.org/latest/guides/modules.html#submitting-changes-to-modules

Or probably even better: you could simply submit and fast-track the manifest update separately, upgrading everyone to the new hal_xtensa even before your Zephyr PR is merged.

It works after cherry-picking everything, thanks!

**Comment by marc-hb** [comment]
> I see that xtensa_build_zephyr.sh is doing west update zephyr hal_xtensa, but maybe is not taking the latest changes?
Check if you have the latest commit from hal_xtensa.

In fact if _you_ run `west update zephyr hal_xtensa` then you should get `xtensa/config/core-isa.h: No such file or directory` right now. Try it, as long as you've committed or `git stash`ed everything in `hal_xtensa` you can't lose anything.

**Pull request closed without merging**